mobile/mobile.go

```diff
@@ -0,0 +1,505 @@
+	"github.com/lightningnetwork/lnd/lnrpc/chainrpc"
+	"github.com/lightningnetwork/lnd/lnrpc/invoicesrpc"
+	"github.com/lightningnetwork/lnd/lnrpc/routerrpc"
+	"github.com/lightningnetwork/lnd/lnrpc/signrpc"
+	"github.com/lightningnetwork/lnd/lnrpc/verrpc"
+	"github.com/lightningnetwork/lnd/lnrpc/walletrpc"
+	"github.com/lightningnetwork/lnd/lnrpc/watchtowerrpc"
+	"github.com/lightningnetwork/lnd/lnrpc/wtclientrpc"
+	"github.com/lightningnetwork/lnd/signal"
+	"google.golang.org/grpc"
+	"gopkg.in/macaroon-bakery.v2/bakery"
+	"gopkg.in/macaroon-bakery.v2/bakery/checkers"
+	"gopkg.in/macaroon.v2"
+)
+
+type mobileClient struct {
+	lndConn *grpc.ClientConn
+
+	statusChecker func() mailbox.ClientStatus
+
+	mac *macaroon.Macaroon
+
+	registry map[string]func(context.Context, *grpc.ClientConn,
+		string, func(string, error))
+}
+
+type stubPackageRegistration func(map[string]func(context.Context,
+	*grpc.ClientConn, string, func(string, error)))
+
+type NativeCallback interface {
+	SendResult(json string)
+}
+
+var (
+	registrations = []stubPackageRegistration{
+		lnrpc.RegisterLightningJSONCallbacks,
+		lnrpc.RegisterStateJSONCallbacks,
+		autopilotrpc.RegisterAutopilotJSONCallbacks,
+		chainrpc.RegisterChainNotifierJSONCallbacks,
+		invoicesrpc.RegisterInvoicesJSONCallbacks,
+		routerrpc.RegisterRouterJSONCallbacks,
+		signrpc.RegisterSignerJSONCallbacks,
+		verrpc.RegisterVersionerJSONCallbacks,
+		walletrpc.RegisterWalletKitJSONCallbacks,
+		watchtowerrpc.RegisterWatchtowerJSONCallbacks,
+		wtclientrpc.RegisterWatchtowerClientJSONCallbacks,
+		looprpc.RegisterSwapClientJSONCallbacks,
+		poolrpc.RegisterTraderJSONCallbacks,
+		frdrpc.RegisterFaradayServerJSONCallbacks,
+	}
+
+	perms = core.GetAllMethodPermissions()
+
+	jsonCBRegex = regexp.MustCompile(`(\w+)\.(\w+)\.(\w+)`)
+
+	m *mobileClient
+
+	registry = make(map[string]func(context.Context,
+		*grpc.ClientConn, string, func(string, error)))
+
+	localPrivCreateCallback  NativeCallback
+	remoteKeyReceiveCallback NativeCallback
+	authDataCallback         NativeCallback
+)
+
+// InitLNC sets up everything required for LNC to run including
+// signal interceptor, logs, and an instance of the mobile client.
+func InitLNC(debugLevel string) {
+	defer func() {
+		if r := recover(); r != nil {
+			fmt.Printf("Recovered in f: %v", r)
+			debug.PrintStack()
+		}
+	}()
+
+	// set debug level to 'info' if not specified
+	if debugLevel == "" {
+		debugLevel = "info"
+	}
+
+	// Hook interceptor for os signals.
+	shutdownInterceptor, err := signal.Intercept()
+	if err != nil {
+		exit(err)
+	}
+
+	logWriter := build.NewRotatingLogWriter()
+	SetupLoggers(logWriter, shutdownInterceptor)
+
+	err = build.ParseAndSetDebugLevels(debugLevel, logWriter)
+	if err != nil {
+		exit(err)
+	}
+
+	m = newMobileClient()
+
+	for _, registration := range registrations {
+		registration(registry)
+	}
+
+	log.Debugf("Mobile client ready for connecting")
+}
+
+func newMobileClient() *mobileClient {
```

Comment:
nit: generally nicer to have a constructor of a struct directly beneath the struct definition